### ISSUE TITLE
istio-proxy left running without app container ever starting - stuck pods

### ISSUE THREAD

**Issue by jmagnusson** [redboxllc/scuttle#34]
Hi,

We sometimes (maybe every 100th run) get logs like this for our CronJob-created pods:

```
2020-06-13T14:15:06.871856039Z scuttle: Logging is now enabled
2020-06-13T14:15:06.871942104Z scuttle: Blocking until envoy starts
2020-06-13T14:15:09.820310817Z scuttle: Blocking finished, envoy has started
2020-06-13T14:15:09.832039127Z scuttle: Received exit signal, exiting
```

When this happens it leaves `istio-proxy` running for some reason. This results in the job never finishing. Which means a new job will never start (because we don't allow more than one running job).

As you can see from the log messages above it seems that scuttle exits pretty much immediately after `envoy has started` is logged.

We're running Scuttle 1.2.3, Istio 1.5 and Kubernetes 1.16.

Looking in main.go I noticed [this comment](https://github.com/redboxllc/scuttle/blob/v1.3.0/main.go#L70):
`// Signal received before the process even started. Let's just exit.`

And that does indeed seem to be the case.

I can't find any relevant istio-proxy logs more than this line: `2020-06-13T14:15:08.986202Z	info	Envoy proxy is ready`

Does anyone have similar experiences? What can be done to fix this?

I'm not fluent in golang so I don't feel comfortable digging further into the code / submitting a PR, but some suggestions maybe to help with debugging these kinds of issues:

1. Split the following if statement into two separate ones: https://github.com/redboxllc/scuttle/blob/v1.3.0/main.go#L43 - One for the `err == nil` case and one for the `errors.Is(err, context.Canceled)` case. Add as much context as possible to the logging.
1. More detailed logging here: https://github.com/redboxllc/scuttle/blob/v1.3.0/main.go#L70 (which signal was received, move comment `Signal received before the process even started. Let's just exit.` into log message)

@linjmeyer What are your thoughts? Do you have some experience in debugging this? (I noticed the same log message in #23)

**Comment by linjmeyer** [comment]
I'll see if I can reproduce, we do have some unit/integration tests and may be able to reproduce this in those.  

The other Issue you linked to (#23) the root cause was that the python script that was running inside the CronJob was not logging anything, including the error it was raising.  So the problem wasn't scuttle itself, but the lack of info in the logs made it trickier to understand that.  However looking at the comment above where that log line is output, the python script never would have started.  I'm a bit confused by that, but the comment came from the original repo Scuttle is forked from so I will need to dig deeper into it.  Maybe it's not even accurate?


In the mean time I will:
* try to reproduce locally
* add better logging
* look into stopping Istio in this case.  If Istio hasn't started up, the call to stop Istio will fail but there is no harm in trying

**Comment by rhavermans** [comment]
Same here, with Java pod.

**Comment by linjmeyer** [comment]
I haven't been able to reproduce so if either of you have any tips/ideas on doing so that would be great.  We use scuttle pretty heavily and haven't seen it yet.  I'm thinking through what would cause Scuttle to receive an exit code prior to it starting the workload app and the only thing that really comes to mind is kubernetes killing the container, but that explanation doesn't make a lot of sense considering Kubernetes would have just started the container.  If you see it happen can you check the kubernetes events for the pod just to see if there is anything interesting in there?

If I can't reproduce it I think the easy workaround could be just having Scuttle send the Istio /quitquitquit call anyway.  Worst case it fails and we see this issue again, best case it stops Istio and K8s will restart everything/allow a new job to run.

**Comment by timschadenberg** [comment]
In the pods that @rhavermans is talking about there are actually no events at all. Below some logging from the istio-proxy with Envoy start parameters.

```
...
2020-07-01T17:45:43.837790Z	info	Envoy command: [-c etc/istio/proxy/envoy-rev0.json --restart-epoch 0 --drain-time-s 45 --parent-shutdown-time-s 60 --service-cluster *** --service-node sidecar~***~***~*** --max-obj-name-len 189 --local-address-ip-version v4 --log-format %Y-%m-%dT%T.%fZ	%l	envoy %n	%v -l warning --component-log-level misc:error --concurrency 2]
2020-07-01T17:45:43.991219Z	warning	envoy config	[bazel-out/k8-opt/bin/external/envoy/source/common/config/_virtual_includes/grpc_stream_lib/common/config/grpc_stream.h:92] StreamAggregatedResources gRPC config stream closed: 14, no healthy upstream
2020-07-01T17:45:43.991292Z	warning	envoy config	[bazel-out/k8-opt/bin/external/envoy/source/common/config/_virtual_includes/grpc_stream_lib/common/config/grpc_stream.h:54] Unable to establish new stream
2020-07-01T17:45:44.004058Z	info	sds	resource:default new connection
2020-07-01T17:45:44.004182Z	info	sds	Skipping waiting for ingress gateway secret
2020-07-01T17:45:44.898472Z	info	cache	Root cert has changed, start rotating root cert for SDS clients
2020-07-01T17:45:44.898539Z	info	cache	GenerateSecret default
2020-07-01T17:45:44.898741Z	info	sds	resource:default pushed key/cert pair to proxy
2020-07-01T17:45:44.991604Z	info	sds	resource:ROOTCA new connection
2020-07-01T17:45:44.991734Z	info	sds	Skipping waiting for ingress gateway secret
2020-07-01T17:45:44.991779Z	info	cache	Loaded root cert from certificate ROOTCA
2020-07-01T17:45:44.991908Z	info	sds	resource:ROOTCA pushed root cert to proxy
2020-07-01T17:45:45.795092Z	info	Envoy proxy is ready
```

```
2020-07-01T17:45:25Z scuttle: Logging is now enabled
2020-07-01T17:45:25Z scuttle: Blocking until Envoy starts
2020-07-01T17:45:45Z scuttle: Blocking finished, Envoy has started
2020-07-01T17:45:45Z scuttle: Received exit signal, exiting
```

**Comment by linjmeyer** [comment]
Hi @timschadenberg, @rhavermans, @jmagnusson 

We just released v1.3.1 to solve this issue.  I was able to reproduce this by running something like `scuttle sleep 9999` and then in another terminal killing the scuttle process with `kill -INT $(pidof scuttle)`.  This sends an interrupt signal from the OS to Scuttle.  If the child process scuttle is wrapping has already started, it passes this signal into that process and the child determines how to react.  If the child process has not been started (for example if scuttle is still waiting on Envoy), then scuttle will trigger the usual sidecar shutdown procedure based on how you have configured scuttle.

In addition to that we adjusted some logs for clarity, including logging whenever a signal is received from the OS (or k8s, docker, etc.) and what the signal is:

When child process was started:
`scuttle: Received signal 'interrupt', passing to child`
or when child was not started:
`scuttle: Received signal 'interrupt', exiting`

If you still see the issue tag me on this thread and I'll re-open, the additional logging around signal handling should help us get some more details.
